mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/blob/BlobStoreDeletedMessageVault.java

```diff
@@ -0,0 +1,93 @@
+ ****************************************************************/
+
+package org.apache.james.vault.blob;
+
+import java.io.InputStream;
+
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.james.blob.api.BlobStore;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.core.User;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.task.Task;
+import org.apache.james.vault.DeletedMessage;
+import org.apache.james.vault.DeletedMessageVault;
+import org.apache.james.vault.metadata.DeletedMessageMetadataVault;
+import org.apache.james.vault.metadata.DeletedMessageWithStorageInformation;
+import org.apache.james.vault.metadata.StorageInformation;
+import org.apache.james.vault.search.Query;
+import org.reactivestreams.Publisher;
+
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class BlobStoreDeletedMessageVault implements DeletedMessageVault {
+    private final DeletedMessageMetadataVault messageMetadataVault;
+    private final BlobStore blobStore;
+    private final BucketNameGenerator nameGenerator;
+
+    public BlobStoreDeletedMessageVault(DeletedMessageMetadataVault messageMetadataVault, BlobStore blobStore, BucketNameGenerator nameGenerator) {
+        this.messageMetadataVault = messageMetadataVault;
+        this.blobStore = blobStore;
+        this.nameGenerator = nameGenerator;
+    }
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage, InputStream mimeMessage) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+        Preconditions.checkNotNull(mimeMessage);
+        BucketName bucketName = nameGenerator.currentBucket();
+        return blobStore.save(bucketName, mimeMessage)
+            .map(blobId -> new StorageInformation(bucketName, blobId))
+            .map(storageInformation -> new DeletedMessageWithStorageInformation(deletedMessage, storageInformation))
+            .flatMap(message -> Mono.from(messageMetadataVault.store(message)))
+            .then();
+    }
+
+    @Override
+    public Publisher<InputStream> loadMimeMessage(User user, MessageId messageId) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+        return Mono.from(messageMetadataVault.retrieveStorageInformation(user, messageId))
+            .map(storageInformation -> blobStore.read(storageInformation.getBucketName(), storageInformation.getBlobId()));
+    }
+
+    @Override
+    public Publisher<DeletedMessage> search(User user, Query query) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(query);
+        return Flux.from(messageMetadataVault.listRelatedBuckets())
+            .concatMap(bucketName -> Flux.from(messageMetadataVault.listMessages(bucketName, user)))
+            .map(DeletedMessageWithStorageInformation::getDeletedMessage);
```

Comment:
did you miss filter() step to be able to select matched deleted messages?

Reply:
We are missing a test in the contract then...